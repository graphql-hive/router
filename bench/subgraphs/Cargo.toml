[package]
name = "subgraphs"
version = "0.0.1"
edition = "2021"

[[bin]]
name = "subgraphs"
path = "main.rs"

[lib]
name = "subgraphs"
path = "lib.rs"

[dependencies]
async-graphql = "7.0.17"
async-graphql-axum = "7.0.17"
axum = "0.8.4"
lazy_static = "1.5.0"
rand = "0.9.1"
tokio = { version = "1.45.1", features = ["macros", "rt-multi-thread"] }
