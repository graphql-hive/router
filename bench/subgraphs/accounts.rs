use async_graphql::{EmptyMutation, EmptySubscription, Object, Schema, SimpleObject, ID};
use lazy_static::lazy_static;

lazy_static! {
    static ref USERS: Vec<User> = vec![
        User {
            id: ID("1".to_string()),
            name: Some("Uri Goldshtein".to_string()),
            username: Some("urigo".to_string()),
            birthday: Some(1234567890),
        },
        User {
            id: ID("2".to_string()),
            name: Some("Dotan Simha".to_string()),
            username: Some("dotansimha".to_string()),
            birthday: Some(1234567890),
        },
        User {
            id: ID("3".to_string()),
            name: Some("Kamil Kisiela".to_string()),
            username: Some("kamilkisiela".to_string()),
            birthday: Some(1234567890),
        },
        User {
            id: ID("4".to_string()),
            name: Some("Arda Tanrikulu".to_string()),
            username: Some("ardatan".to_string()),
            birthday: Some(1234567890),
        },
        User {
            id: ID("5".to_string()),
            name: Some("Gil Gardosh".to_string()),
            username: Some("gilgardosh".to_string()),
            birthday: Some(1234567890),
        },
        User {
            id: ID("6".to_string()),
            name: Some("Laurin Quast".to_string()),
            username: Some("laurin".to_string()),
            birthday: Some(1234567890),
        }
    ];
}

#[derive(SimpleObject, Clone)]
pub struct User {
    id: ID,
    name: Option<String>,
    username: Option<String>,
    birthday: Option<i32>,
}

impl User {
    fn me() -> User {
        USERS[0].clone()
    }
}

pub struct Query;

#[Object(extends = true)]
impl Query {
    async fn me(&self) -> Option<User> {
        Some(User::me())
    }

    async fn user(&self, id: ID) -> Option<User> {
        USERS.iter().find(|user| user.id == id).cloned()
    }

    async fn users(&self) -> Option<Vec<Option<User>>> {
        Some(USERS.iter().map(|user| Some(user.clone())).collect())
    }

    #[graphql(entity)]
    async fn find_user_by_id(&self, id: ID) -> User {
        USERS.iter().find(|user| user.id == id).cloned().unwrap()
    }
}

pub fn get_subgraph() -> Schema<Query, EmptyMutation, EmptySubscription> {
    Schema::build(Query, EmptyMutation, EmptySubscription)
        .enable_federation()
        .finish()
}
