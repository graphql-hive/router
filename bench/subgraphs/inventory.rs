use async_graphql::{
    ComplexObject, EmptyMutation, EmptySubscription, Object, Schema, SimpleObject, ID,
};
use lazy_static::lazy_static;

lazy_static! {
    static ref INVENTORY: Vec<Product> = vec![
        Product {
            upc: "1".to_string(),
            in_stock: Some(true),
            price: None,
            weight: None,
        },
        Product {
            upc: "2".to_string(),
            in_stock: Some(false),
            price: None,
            weight: None,
        },
        Product {
            upc: "3".to_string(),
            in_stock: Some(false),
            price: None,
            weight: None,
        },
        Product {
            upc: "4".to_string(),
            in_stock: Some(false),
            price: None,
            weight: None,
        },
        Product {
            upc: "5".to_string(),
            in_stock: Some(true),
            price: None,
            weight: None,
        },
        Product {
            upc: "6".to_string(),
            in_stock: Some(true),
            price: None,
            weight: None,
        },
        Product {
            upc: "7".to_string(),
            in_stock: Some(true),
            price: None,
            weight: None,
        },
        Product {
            upc: "8".to_string(),
            in_stock: Some(false),
            price: None,
            weight: None,
        },
        Product {
            upc: "9".to_string(),
            in_stock: Some(true),
            price: None,
            weight: None,
        }
    ];
}

#[derive(SimpleObject, Clone)]
#[graphql(extends, complex)]
pub struct Product {
    #[graphql(external)]
    upc: String,
    #[graphql(external)]
    weight: Option<i64>,
    #[graphql(external)]
    price: Option<i64>,
    in_stock: Option<bool>,
}

#[ComplexObject]
impl Product {
    #[graphql(requires = "price weight")]
    pub async fn shipping_estimate(&self) -> Option<i64> {
        if let Some(price) = self.price {
            if price > 1000 {
                return Some(0);
            }

            if let Some(weight) = self.weight {
                return Some(weight / 2);
            }
        }

        None
    }
}

pub struct Query;

#[Object(extends = true)]
impl Query {
    #[graphql(entity)]
    async fn find_product_by_id(&self, upc: ID) -> Product {
        INVENTORY
            .iter()
            .find(|product| product.upc == upc.to_string())
            .unwrap()
            .clone()
    }
}

pub fn get_subgraph() -> Schema<Query, EmptyMutation, EmptySubscription> {
    Schema::build(Query, EmptyMutation, EmptySubscription)
        .enable_federation()
        .finish()
}
