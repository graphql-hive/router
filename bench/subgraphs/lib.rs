pub mod accounts;
pub mod inventory;
pub mod products;
pub mod reviews;
