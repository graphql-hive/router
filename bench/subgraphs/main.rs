use async_graphql_axum::GraphQL;
use axum::{
    extract::Request,
    middleware::{self, Next},
    response::Response,
    routing::post_service,
    Router,
};
use std::env::var;
use tokio::net::TcpListener;

mod accounts;
mod inventory;
mod products;
mod reviews;

extern crate lazy_static;

async fn delay_middleware(req: Request, next: Next) -> Response {
    let delay_ms: Option<u64> = std::env::var("SUBGRAPH_DELAY_MS")
        .ok()
        .and_then(|s| s.parse().ok())
        .filter(|d| *d != 0);

    if let Some(delay_ms) = delay_ms {
        tokio::time::sleep(tokio::time::Duration::from_millis(delay_ms)).await;
    }

    next.run(req).await
}

#[tokio::main]
async fn main() {
    let host = var("HOST").unwrap_or("0.0.0.0".to_owned());
    let port = var("PORT").unwrap_or("4200".to_owned());

    let app = Router::new()
        .route(
            "/accounts",
            post_service(GraphQL::new(accounts::get_subgraph())),
        )
        .route(
            "/inventory",
            post_service(GraphQL::new(inventory::get_subgraph())),
        )
        .route(
            "/products",
            post_service(GraphQL::new(products::get_subgraph())),
        )
        .route(
            "/reviews",
            post_service(GraphQL::new(reviews::get_subgraph())),
        )
        .route_layer(middleware::from_fn(delay_middleware));

    println!("Starting server on http://{}:{}", host, port);

    axum::serve(
        TcpListener::bind(&format!("{}:{}", host, port))
            .await
            .unwrap(),
        app,
    )
    .await
    .unwrap();
}
