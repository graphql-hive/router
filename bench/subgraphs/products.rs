use async_graphql::{EmptyMutation, EmptySubscription, Object, Schema, SimpleObject, ID};
use lazy_static::lazy_static;

lazy_static! {
    static ref PRODUCTS: Vec<Product> = vec![
        Product {
            upc: "1".to_string(),
            name: Some("Table".to_string()),
            price: Some(899),
            weight: Some(100),
        },
        Product {
            upc: "2".to_string(),
            name: Some("Couch".to_string()),
            price: Some(1299),
            weight: Some(1000),
        },
        Product {
            upc: "3".to_string(),
            name: Some("Glass".to_string()),
            price: Some(15),
            weight: Some(20),
        },
        Product {
            upc: "4".to_string(),
            name: Some("Chair".to_string()),
            price: Some(499),
            weight: Some(100),
        },
        Product {
            upc: "5".to_string(),
            name: Some("TV".to_string()),
            price: Some(1299),
            weight: Some(1000),
        },
        Product {
            upc: "6".to_string(),
            name: Some("Lamp".to_string()),
            price: Some(6999),
            weight: Some(300),
        },
        Product {
            upc: "7".to_string(),
            name: Some("Grill".to_string()),
            price: Some(3999),
            weight: Some(2000),
        },
        Product {
            upc: "8".to_string(),
            name: Some("Fridge".to_string()),
            price: Some(100000),
            weight: Some(6000),
        },
        Product {
            upc: "9".to_string(),
            name: Some("Sofa".to_string()),
            price: Some(9999),
            weight: Some(800),
        }
    ];
}
#[derive(SimpleObject, Clone)]
pub struct Product {
    upc: String,
    name: Option<String>,
    price: Option<i64>,
    weight: Option<i64>,
}

pub struct Query;

#[Object(extends = true)]
impl Query {
    async fn top_products(
        &self,
        #[graphql(default = 5)] first: Option<i32>,
    ) -> Option<Vec<Option<Product>>> {
        Some(
            PRODUCTS
                .iter()
                .take(first.unwrap_or(5) as usize)
                .map(|product| Some(product.clone()))
                .collect(),
        )
    }

    #[graphql(entity)]
    async fn find_product_by_upc(&self, upc: ID) -> Product {
        PRODUCTS
            .iter()
            .find(|product| product.upc == upc.as_str())
            .unwrap()
            .clone()
    }
}

pub fn get_subgraph() -> Schema<Query, EmptyMutation, EmptySubscription> {
    Schema::build(Query, EmptyMutation, EmptySubscription)
        .enable_federation()
        .finish()
}
