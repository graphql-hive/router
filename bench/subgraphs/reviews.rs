use async_graphql::{
    ComplexObject, EmptyMutation, EmptySubscription, Object, Schema, SimpleObject, ID,
};
use lazy_static::lazy_static;

lazy_static! {
    static ref REVIEWS: Vec<Review> = vec![
        Review {
            id: ID("1".to_string()),
            body: Some("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.".to_string()),
            product: Some(Product {
                upc: "1".to_string()
            })
        },
        Review {
            id: ID("2".to_string()),
            body: Some("Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugi".to_string()),
            product: Some(Product {
                upc: "1".to_string()
            })
        },
        Review {
            id: ID("3".to_string()),
            body: Some("sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem.".to_string()),
            product: Some(Product {
                upc: "1".to_string()
            })
        },
        Review {
            id: ID("4".to_string()),
            body: Some("Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem".to_string()),
            product: Some(Product {
                upc: "1".to_string()
            })
        },
        Review {
            id: ID("5".to_string()),
            body: Some("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.".to_string()),
            product: Some(Product {
                upc: "2".to_string()
            })
        },
        Review {
            id: ID("6".to_string()),
            body: Some("Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugi".to_string()),
            product: Some(Product {
                upc: "2".to_string()
            })
        },
        Review {
            id: ID("7".to_string()),
            body: Some("sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem.".to_string()),
            product: Some(Product {
                upc: "2".to_string()
            })
        },
        Review {
            id: ID("8".to_string()),
            body: Some("Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem".to_string()),
            product: Some(Product {
                upc: "2".to_string()
            })
        },
        Review {
            id: ID("9".to_string()),
            body: Some("Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem".to_string()),
            product: Some(Product {
                upc: "3".to_string()
            })
        },
        Review {
            id: ID("10".to_string()),
            body: Some("Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem".to_string()),
            product: Some(Product {
                upc: "4".to_string()
            })
        },
        Review {
            id: ID("11".to_string()),
            body: Some("At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores et quas molestias excepturi sint occaecati cupiditate non provident, similique sunt in culpa qui officia deserunt mollitia animi, id est laborum et dolorum fuga. Et harum quidem rerum facilis est et expedita distinctio. Nam libero tempore, cum soluta nobis est eligendi optio cumque nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.".to_string()),
            product: Some(Product {
                upc: "4".to_string()
            })
        }
    ];
}

#[derive(SimpleObject, Clone)]
#[graphql(complex)]
pub struct Review {
    pub id: ID,
    pub body: Option<String>,
    pub product: Option<Product>,
}

#[ComplexObject]
impl Review {
    #[graphql(provides = "username")]
    pub async fn author(&self) -> Option<User> {
        Some(User {
            id: "1".into(),
            username: Some("urigo".to_string()),
            reviews: Some(REVIEWS[0..2].iter().map(|r| Some(r.clone())).collect()),
        })
    }
}

#[derive(SimpleObject)]
#[graphql(extends)]
pub struct User {
    #[graphql(external)]
    id: ID,
    #[graphql(external)]
    username: Option<String>,
    reviews: Option<Vec<Option<Review>>>,
}

#[derive(SimpleObject, Clone)]
#[graphql(extends, complex)]
pub struct Product {
    #[graphql(external)]
    upc: String,
}

#[ComplexObject]
impl Product {
    pub async fn reviews(&self) -> Option<Vec<Option<Review>>> {
        let relevant = REVIEWS
            .iter()
            .filter(|r| {
                if let Some(product) = &r.product {
                    product.upc == self.upc
                } else {
                    false
                }
            })
            .map(|r| Some(r.clone()))
            .collect();

        Some(relevant)
    }
}

pub struct Query;

#[Object(extends = true)]
impl Query {
    #[graphql(entity)]
    async fn find_review_by_id(&self, id: ID) -> Review {
        REVIEWS.iter().find(|r| r.id == id).unwrap().clone()
    }

    #[graphql(entity)]
    async fn find_user_by_id(&self, id: ID) -> User {
        User {
            id,
            username: Some("user".to_string()),
            reviews: Some(REVIEWS[0..2].iter().map(|r| Some(r.clone())).collect()),
        }
    }

    #[graphql(entity)]
    async fn find_product_by_id(&self, upc: ID) -> Product {
        Product {
            upc: upc.to_string(),
        }
    }
}

pub fn get_subgraph() -> Schema<Query, EmptyMutation, EmptySubscription> {
    Schema::build(Query, EmptyMutation, EmptySubscription)
        .enable_federation()
        .finish()
}
