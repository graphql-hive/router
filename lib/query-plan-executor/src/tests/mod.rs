use std::{collections::HashMap, sync::Arc};

use subgraphs::accounts;

use crate::{executors::async_graphql::AsyncGraphQLExecutor, SubgraphExecutorMap};

#[test]
fn query_executor_pipeline_locally() {
    tokio_test::block_on(async {
        let operation_path = "../../bench/operation.graphql";
        let supergraph_sdl = std::fs::read_to_string("../../bench/supergraph.graphql")
            .expect("Unable to read input file");
        let parsed_schema = query_planner::utils::parsing::parse_schema(&supergraph_sdl);
        let planner = query_planner::planner::Planner::new_from_supergraph(&parsed_schema)
            .expect("Failed to create planner from supergraph");
        let parsed_document = query_planner::utils::parsing::parse_operation(
            &std::fs::read_to_string(operation_path).expect("Unable to read input file"),
        );
        let normalized_document = query_planner::ast::normalization::normalize_operation(
            &planner.supergraph,
            &parsed_document,
            None,
        )
        .expect("Failed to normalize operation");
        let normalized_operation = normalized_document.executable_operation();
        let query_plan = planner
            .plan_from_normalized_operation(normalized_operation)
            .expect("Failed to create query plan");
        let schema_metadata =
            crate::schema_metadata::SchemaWithMetadata::schema_metadata(&planner.consumer_schema);
        let mut subgraph_executor_map: SubgraphExecutorMap = HashMap::new(); // No subgraphs in this test
        let accounts = accounts::get_subgraph();
        let inventory = subgraphs::inventory::get_subgraph();
        let products = subgraphs::products::get_subgraph();
        let reviews = subgraphs::reviews::get_subgraph();
        subgraph_executor_map.insert(
            "accounts".to_string(),
            Arc::new(Box::new(AsyncGraphQLExecutor::new(accounts))),
        );
        subgraph_executor_map.insert(
            "inventory".to_string(),
            Arc::new(Box::new(AsyncGraphQLExecutor::new(inventory))),
        );
        subgraph_executor_map.insert(
            "products".to_string(),
            Arc::new(Box::new(AsyncGraphQLExecutor::new(products))),
        );
        subgraph_executor_map.insert(
            "reviews".to_string(),
            Arc::new(Box::new(AsyncGraphQLExecutor::new(reviews))),
        );
        let result = crate::execute_query_plan(
            &query_plan,
            &subgraph_executor_map,
            &None,
            &schema_metadata,
            normalized_operation,
            false,
        )
        .await;
        insta::assert_snapshot!(format!(
            "{}",
            serde_json::to_string_pretty(&result).unwrap_or_default()
        ));
    });
}
