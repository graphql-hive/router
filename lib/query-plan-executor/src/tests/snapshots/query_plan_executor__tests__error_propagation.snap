---
source: lib/query-plan-executor/src/tests/mod.rs
expression: result_data
---
{"movie1":{"__typename":"Movie","director":{"id":"1","name":"Christopher Nolan"},"id":"1","name":"Inception"},"movie2":{"__typename":"Movie","id":"2","name":"Interstellar"}}
